File: src-tauri/src/export/spring.rs
use crate::export::types::Easing;

pub const PREFIX: &str = "spring(";

pub const STIFFNESS: (f32, f32) = (1.0, 2000.0);
pub const DAMPING: (f32, f32) = (0.0, 200.0);
pub const MASS: (f32, f32) = (0.1, 10.0);

const ZETA_MIN: f32 = 0.05;

const LN_EPS: f32 = 6.907_755;

const CRIT: f32 = 1e-3;

fn zeta(stiffness: f32, damping: f32, mass: f32) -> f32 {
    let k = stiffness.clamp(STIFFNESS.0, STIFFNESS.1);
    let (c, m) = (
        damping.clamp(DAMPING.0, DAMPING.1),
        mass.clamp(MASS.0, MASS.1),
    );
    (c / (2.0 * (k * m).sqrt())).max(ZETA_MIN)
}

fn settle_u(z: f32) -> f32 {
    LN_EPS
        / if z <= 1.0 {
            z
        } else {
            z - (z * z - 1.0).max(0.0).sqrt()
        }
}

fn resp(z: f32, u: f32) -> f32 {
    if (z - 1.0).abs() <= CRIT {
        return 1.0 - (1.0 + u) * (-u).exp();
    }
    if z < 1.0 {
        let d = (1.0 - z * z).sqrt();
        return 1.0 - (-z * u).exp() * ((d * u).cos() + z / d * (d * u).sin());
    }
    let s = (z * z - 1.0).max(0.0).sqrt();
    let (r1, r2) = (-(z - s), -(z + s));
    1.0 - (r2 * (r1 * u).exp() - r1 * (r2 * u).exp()) / (r2 - r1)
}

pub fn spring(stiffness: f32, damping: f32, mass: f32, p: f32) -> f32 {
    if !p.is_finite() || p <= 0.0 {
        return 0.0;
    }
    if p >= 1.0 {
        return 1.0;
    }
    let (z, us) = {
        let z = zeta(stiffness, damping, mass);
        (z, settle_u(z))
    };
    resp(z, p * us) + p * (1.0 - resp(z, us))
}

pub fn eval(e: Easing, p: f32) -> f32 {
    match e {
        Easing::Spring {
            stiffness,
            damping,
            mass,
        } => spring(stiffness, damping, mass, p),
        _ => p,
    }
}

pub fn parse_spring(s: &str) -> Option<(f32, f32, f32)> {
    let inner = s.trim().strip_prefix(PREFIX)?.strip_suffix(')')?;
    let mut v = [0f32; 3];
    let mut n = 0;
    for part in inner.split(',') {
        if n == 3 {
            return None;
        }
        v[n] = part.trim().parse::<f32>().ok()?;
        if !v[n].is_finite() {
            return None;
        }
        n += 1;
    }
    if n < 2 {
        return None;
    }
    if n == 2 {
        v[2] = 1.0;
    }
    Some((
        v[0].clamp(STIFFNESS.0, STIFFNESS.1),
        v[1].clamp(DAMPING.0, DAMPING.1),
        v[2].clamp(MASS.0, MASS.1),
    ))
}

pub fn format_spring(stiffness: f32, damping: f32, mass: f32) -> String {
    format!("spring({stiffness:.3},{damping:.3},{mass:.3})")
}

#[cfg(test)]
#[path = "spring_tests.rs"]
mod tests;

File: src/shared/math/spring.ts
export const SPRING_RANGE = { stiffness: [1, 2000], damping: [0, 200], mass: [0.1, 10] } as const;

export const SPRING_DEFAULT: [number, number, number] = [100, 10, 1];

const ZETA_MIN = 0.05;
const LN_EPS = 6.907755;
const CRIT = 1e-3;
const clamp = (v: number, lo: number, hi: number) => Math.min(hi, Math.max(lo, v));

function zeta(stiffness: number, damping: number, mass: number): number {
  const k = clamp(stiffness, 1, 2000),
    c = clamp(damping, 0, 200),
    m = clamp(mass, 0.1, 10);
  return Math.max(ZETA_MIN, c / (2 * Math.sqrt(k * m)));
}

const settleU = (z: number) => LN_EPS / (z <= 1 ? z : z - Math.sqrt(Math.max(0, z * z - 1)));

function resp(z: number, u: number): number {
  if (Math.abs(z - 1) <= CRIT) return 1 - (1 + u) * Math.exp(-u);
  if (z < 1) {
    const d = Math.sqrt(1 - z * z);
    return 1 - Math.exp(-z * u) * (Math.cos(d * u) + (z / d) * Math.sin(d * u));
  }
  const s = Math.sqrt(Math.max(0, z * z - 1));
  const r1 = -(z - s),
    r2 = -(z + s);
  return 1 - (r2 * Math.exp(r1 * u) - r1 * Math.exp(r2 * u)) / (r2 - r1);
}

export function spring(stiffness: number, damping: number, mass: number, p: number): number {
  if (!Number.isFinite(p) || p <= 0) return 0;
  if (p >= 1) return 1;
  const z = zeta(stiffness, damping, mass),
    us = settleU(z);
  return resp(z, p * us) + p * (1 - resp(z, us));
}

export function parseSpring(s: string | undefined | null): [number, number, number] | null {
  const m = typeof s === "string" ? s.trim().match(/^spring\(([^()]*)\)$/) : null;
  if (!m) return null;
  const parts = m[1].split(",");
  if (parts.length < 2 || parts.length > 3) return null;
  const v = parts.map((x) => Number(x.trim()));
  if (v.some((n) => !Number.isFinite(n)) || parts.some((x) => x.trim() === "")) return null;
  return [clamp(v[0], 1, 2000), clamp(v[1], 0, 200), clamp(v[2] ?? 1, 0.1, 10)];
}

export const formatSpring = (stiffness: number, damping: number, mass = 1) =>
  `spring(${stiffness.toFixed(3)},${damping.toFixed(3)},${mass.toFixed(3)})`;

export function springOf(easing: string | undefined | null): [number, number, number] | null {
  if (easing === "spring") return [...SPRING_DEFAULT];
  return parseSpring(easing);
}
